### ISSUE TITLE
Add enum support when generating entities from database schema

### ISSUE THREAD

**Issue by kevinh-canva** [mikro-orm/mikro-orm#2598]
**Is your feature request related to a problem? Please describe.**
I haven't tried with other databases, but for Postgres when generating entities `.ts` files from an existing database schema, the enum type is swallowed and the field has type `string` instead, so we lose enum type safety. We could manually add enum definition to the generated file, but the next time we need to re-generate the file due to a schema change, it will overwrite the enum definition.

**Describe the solution you'd like**
I will raise a PR to generate the enum definition in the same entity .ts file and use that for the field.

**Describe alternatives you've considered**
- Manually add enum definition to generated file -> risk the addition being overwritten the next time the file is re-generated

**Additional context**
@B4nan Would love to know if you have looked into this before and if it's feasible at all, since I'm not too familiar with knex which I guess is what we use to get the entity information? Would we be able to tell if a particular field is of type enum and if so what's the definition?

**Comment by B4nan** [comment]
Are you talking about native pg enums? MikroORM does not use those, enums are represented via text fields with check constraints.

I am not opposed by introducing native pg enums support too, if you would contribut it, but it will be much more work than fixing the entity generator part. Note that from what I know, native pg enums should be supported on runtime - they are not supported in schema generator, so this would include:

- handling this new format when creating tables
- diffing of enums
- diffing of text format with check constraints vs native pg enums (in both way)
- all of this would have to work optionally (based on some flag on `@Enum()` decorator)
- enum arrays would have to address this somehow too i guess
- and then finally we can talk about entity generator :]

Overall quite a lot of work for rather small benefit, as using text fields with check constrains seems to be good enough solution. But I can understand that is not valid reason if you already have your existing database and you want to map it in the ORM.

> since I'm not too familiar with knex which I guess is what we use to get the entity information?

Most of the metadata is taken directly from information_schema, knex does not help much with advanced cases.

**Comment by kevinh-canva** [comment]
> 

Yep, this is my first time contributing to an ORM tbh, so the above list seems a fair bit more than what a newbie like me would expect to be able to do :D 

I think it's beneficial to limit the scope of this issue. So right now when I generate the entities from schema using entity generator, even if we use check constraint, the `.ts` code property is generated with string type. So the main issue I want to fix is to generate this as an enum type. 

Let's say I have a `task` table, with `status` enum column on it with 3 values `OPEN, IN_PROGRESS, DONE`. It would be nice if the resulting entity file looks something like this:
```
@Entity()
export class Task {
   ...

   @Enum({ items: () => TaskStatus})
   status: TaskStatus = TaskStatus.OPEN;
}

export enum TaskStatus {
   OPEN = 'OPEN',
   IN_PROGRESS = 'IN_PROGRESS',
   DONE = 'DONE',
}
```

Am I right that just the above change would require a change to entity generator? 

About supporting native pg enum, I agree it's a nice to have. It might be a bit better storage-wise compared to storing text since each value would be represented as 4 bytes, but enum labels are normally short anyway so the savings is not that much. I would be keen if we could just add one-way support for it. For example, my workflow is I always consider the database schema to be the source of truth, so I would handcraft the SQL for creation/migration first, apply the change on y DB, and then run the EntityGenerator to apply the changes to `.ts` entities files. If you are not happy, I could remove those bit from the PR and keep it simple and just support check constraints for now :)

**Comment by B4nan** [comment]
Yeah it definitely does not have to be a single PR, doing it in multiple steps will make everything easier, especially reviewing.

> Am I right that just the above change would require a change to only the entity generator pretty much?

Not really, you will need changes in knex packages too, as you need to get the information about existing enums from the database and that is not implemented (not for native pg enums). But if we are talking about supporting the enums as check constraints (so the way they work now), that should be probably just about the entity generator package.
